onnxruntime/core/providers/webgpu/tensor/concat.cc

```diff
@@ -125,12 +125,84 @@ Status Concat::ComputeInternal(ComputeContext& context) const {
   }
 
   size_t non_empty_input_count = sizes_in_concat_axis.size();
-
   if (non_empty_input_count + 1 > context.DeviceLimits().maxStorageBuffersPerShaderStage) {
-    // TODO: support when input_count + 1 > maxStorageBuffersPerShaderStage, by raising the limit or run the program in multiple passes.
-    return ORT_MAKE_STATUS(ONNXRUNTIME, FAIL, "The number of storage buffer (input=",
-                           input_count, ", output=1) exceeds the limit (",
-                           context.DeviceLimits().maxStorageBuffersPerShaderStage, ") of the device.");
+    // Run concat operation in multiple passes when storage buffer limit is exceeded
+    LOGS_DEFAULT(WARNING) << "Storage buffer limit exceeded for Concat. "
+                             "Running operation in multiple passes.";
+    size_t max_inputs_per_pass = context.DeviceLimits().maxStorageBuffersPerShaderStage - 1;  // Reserve 1 for output
+
+    std::vector<Tensor> intermediate_tensors;
+    std::vector<const Tensor*> current_inputs;
+
+    for (int i = 0; i < input_count; ++i) {
+      current_inputs.push_back(prepare.inputs[i].tensor);
+    }
+
+    // Process inputs in passes until we have few enough to process in one final pass
+    while (current_inputs.size() > max_inputs_per_pass) {
+      // Calculate output shape for this pass
+      auto pass_shape = current_inputs[0]->Shape();
+      int64_t pass_concat_axis_size = 0;
+      for (size_t i = 0; i < max_inputs_per_pass; ++i) {
+        pass_concat_axis_size += current_inputs[i]->Shape()[axis];
+      }
+      pass_shape[axis] = pass_concat_axis_size;
+
+      // Create intermediate tensor to store result of this pass
+      auto intermediate = context.CreateGPUTensor(current_inputs[0]->DataType(), std::move(pass_shape));
```

Comment:
Is it possible to always use `prepare.output_tensor` as output so that creating intermediate buffers can be avoided?